File: capy/src/e2e-cfun/js/dev/capylang/test/PrimitiveBackedTypes.test.js
const test = require('node:test');
const { assert, assertError, assertSuccess, fs, generatedModule, modulePath } = require('./_support');

test('PrimitiveBackedTypes', () => {
    const primitiveBackedTypes = generatedModule('dev/capylang/test/PrimitiveBackedTypes.js');
    assert.equal(primitiveBackedTypes.rawUserId(7), 7);
    assert.equal(primitiveBackedTypes.unwrapUserId(7), 7);
    assert.equal(primitiveBackedTypes.unwrapRawUserId(7), 7);
    assert.equal(primitiveBackedTypes.plusUserIds(2, 3), 5);
    assert.equal(primitiveBackedTypes.addUserIds(4, 5), 9);
    assert.equal(primitiveBackedTypes.incrementUserId(6), 7);
    assert.equal(primitiveBackedTypes.stringAt('abc', 1), 'b');
    assert.equal(primitiveBackedTypes.stringGet('abc', 2), 'c');
    assert.equal(primitiveBackedTypes.unwrapScore(primitiveBackedTypes.scoreOf(11)), 11);
    assert.equal(primitiveBackedTypes.rawToken('abc'), 'abc');
    assert.equal(primitiveBackedTypes.unwrapToken('abc'), 'abc');
    assert.equal(primitiveBackedTypes.passTokenToString('abc'), 'abc_suffix');
    assert.equal(assertSuccess(primitiveBackedTypes.newUserId(7)), 7);
    assert.equal(assertError(primitiveBackedTypes.newUserId(0)), 'bad user id');
    assert.deepEqual(primitiveBackedTypes.__capybaraPrimitiveTypes.user_id, {
        cfunType: '/dev/capylang/test/PrimitiveBackedTypes.user_id',
        backingType: 'int',
    });
    assert.deepEqual(primitiveBackedTypes.__capybaraPrimitiveTypes.score, {
        cfunType: '/dev/capylang/test/PrimitiveBackedTypes.score',
        backingType: 'int',
    });
    assert.deepEqual(primitiveBackedTypes.__capybaraPrimitiveTypes.token, {
        cfunType: '/dev/capylang/test/PrimitiveBackedTypes.token',
        backingType: 'String',
    });

    const generated = fs.readFileSync(modulePath('dev/capylang/test/PrimitiveBackedTypes.js'), 'utf8');
    assert.match(generated, /function addUserIds\(left, right\) \{\s+return plus__op_plus__user_id__user_id\(left, right\);/);
    assert.doesNotMatch(generated, /class Token/);
});

File: capy/src/e2e-cfun/py/dev/capylang/test/primitive_backed_types_test.py
import os
import pathlib
import sys
import unittest


generated_dir = pathlib.Path(os.environ["CAPY_E2E_CFUN_PY_GENERATED_DIR"])
sys.path.insert(0, str(generated_dir))

import dev.capylang.test.PrimitiveBackedTypes as primitive_backed_types


class PrimitiveBackedTypesPythonE2ETest(unittest.TestCase):
    def test_primitive_backed_types_execute_as_ints(self):
        self.assertEqual(primitive_backed_types.rawUserId(7), 7)
        self.assertEqual(primitive_backed_types.unwrapUserId(7), 7)
        self.assertEqual(primitive_backed_types.unwrapRawUserId(7), 7)
        self.assertEqual(primitive_backed_types.plusUserIds(2, 3), 5)
        self.assertEqual(primitive_backed_types.addUserIds(4, 5), 9)
        self.assertEqual(primitive_backed_types.incrementUserId(6), 7)
        self.assertEqual(primitive_backed_types.stringAt("abc", 1), "b")
        self.assertEqual(primitive_backed_types.stringGet("abc", 2), "c")
        self.assertEqual(primitive_backed_types.unwrapScore(primitive_backed_types.scoreOf(11)), 11)
        primitive_types = getattr(primitive_backed_types, "__capybaraPrimitiveTypes")
        self.assertEqual(
            primitive_types["user_id"],
            {"cfunType": "/dev/capylang/test/PrimitiveBackedTypes.user_id", "backingType": "int"},
        )
        self.assertEqual(
            primitive_types["score"],
            {"cfunType": "/dev/capylang/test/PrimitiveBackedTypes.score", "backingType": "int"},
        )
